main.go

```diff
@@ -2,203 +2,177 @@ package main
 
 import (
 	"context"
-	"encoding/base64"
 	"errors"
 	"fmt"
 	"net/http"
-	"net/http/httputil"
-	"net/url"
 	"os"
 	"os/signal"
-	"strconv"
+	"path/filepath"
 	"syscall"
 	"time"
 
+	"golang.org/x/sync/errgroup"
+
 	"github.com/go-logr/logr"
 	"github.com/go-logr/zapr"
-	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	flag "github.com/spf13/pflag"
 	"go.uber.org/zap"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
 
 	"github.com/xenitab/azdo-proxy/pkg/auth"
 	"github.com/xenitab/azdo-proxy/pkg/config"
+	"github.com/xenitab/azdo-proxy/pkg/server"
+	"github.com/xenitab/azdo-proxy/pkg/token"
 )
 
 var (
-	port       int
-	configPath string
+	port           string
+	metricsPort    string
+	kubeconfigPath string
+	configPath     string
 )
 
 func init() {
-	flag.IntVar(&port, "port", 8080, "port to bind server to.")
+	flag.StringVar(&port, "port", ":8080", "port to bind proxy server to.")
+	flag.StringVar(&metricsPort, "metrics-port", ":9090", "port to bind metrics endpoint to.")
+	flag.StringVar(&kubeconfigPath, "kubeconfig", "", "absolute path to the kubeconfig file.")
 	flag.StringVar(&configPath, "config", "/var/config.json", "path to configuration file.")
 	flag.Parse()
 }
 
 func main() {
-	// Initiate logs
-	var log logr.Logger
 	zapLog, err := zap.NewProduction()
 	if err != nil {
 		panic(fmt.Sprintf("who watches the watchmen (%v)?", err))
 	}
-	log = zapr.NewLogger(zapLog)
-	setupLog := log.WithName("setup")
-
-	// Read configuration
-	setupLog.Info("Reading configuration file", "path", configPath)
-	config, err := config.LoadConfigurationFromPath(configPath)
-	if err != nil {
-		setupLog.Error(err, "Failed loading configuration")
-		os.Exit(1)
+	logger := zapr.NewLogger(zapLog)
+	if err := run(logger, configPath, metricsPort); err != nil {
+		logger.WithName("main").Error(err, "run error")
 	}
-	remote, err := url.Parse("https://" + config.Domain)
+}
+
+func run(logger logr.Logger, configPath string, metricsPort string) error {
+	mainLogger := logger.WithName("main")
+	mainLogger.Info("read configuration from", "path", configPath)
+	mainLogger.Info("serve metrics", "port", metricsPort, "endpoint", "/metrics")
+
+	authz, err := getAutorization(configPath)
 	if err != nil {
-		setupLog.Error(err, "Invalid Azure DevOps domain")
-		os.Exit(1)
+		return err
 	}
-	auth, err := auth.GenerateAuthorization(*config)
+	client, err := getKubernetesClient(kubeconfigPath)
 	if err != nil {
-		setupLog.Error(err, "Could not generate authorization")
-		os.Exit(1)
+		return fmt.Errorf("invalid kubernetes client: %w", err)
 	}
 
-	// Signal handler
-	done := make(chan os.Signal, 1)
-	signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+	ctx := context.Background()
+	ctx = logr.NewContext(ctx, logger)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	g, ctx := errgroup.WithContext(ctx)
```

Comment:
This is a great change, I really like it. 👍👍👍🥇

Reply:
:+1: